tests/src/opaque_test_driver.c

```diff
@@ -0,0 +1,338 @@
+                                        size_t *out_length )
+{
+    OPQTD_VALIDATE_RET( attributes != NULL );
+    OPQTD_VALIDATE_RET( in != NULL );
+    OPQTD_VALIDATE_RET( out != NULL );
+    OPQTD_VALIDATE_RET( out_length != NULL );
+
+    (void) attributes;
+
+    if( in_length <= OPAQUE_TEST_DRIVER_KEYHEADER_SIZE )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    if( in_length - OPAQUE_TEST_DRIVER_KEYHEADER_SIZE > out_size )
+        return( PSA_ERROR_BUFFER_TOO_SMALL );
+
+    *out_length = in_length - OPAQUE_TEST_DRIVER_KEYHEADER_SIZE;
+    rot13( in + OPAQUE_TEST_DRIVER_KEYHEADER_SIZE, *out_length, out );
+
+    return( PSA_SUCCESS );
+}
+
+psa_status_t opaque_test_driver_generate_key(
+                                        const psa_key_attributes_t *attributes,
+                                        uint8_t *key,
+                                        size_t key_size,
+                                        size_t *key_length )
+{
+    psa_status_t status;
+    uint8_t key_buffer[32];
+    size_t key_buffer_length;
+
+    OPQTD_VALIDATE_RET( attributes != NULL );
+    OPQTD_VALIDATE_RET( key != NULL );
+    OPQTD_VALIDATE_RET( key_length != NULL );
+
+    if( psa_get_key_bits( attributes ) == 0 )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    if( psa_get_key_type( attributes ) != PSA_KEY_TYPE_AES )
+        return( PSA_ERROR_NOT_SUPPORTED );
+
+    if( ( psa_get_key_bits( attributes ) != 128 ) &&      // AES-128
+        ( psa_get_key_bits( attributes ) != 192 ) &&      // AES-192
+        ( psa_get_key_bits( attributes ) != 256 ) )       // AES-256
+        return( PSA_ERROR_NOT_SUPPORTED );
+
+    if( OPAQUE_TEST_DRIVER_KEYHEADER_SIZE +
+        PSA_BITS_TO_BYTES( psa_get_key_bits( attributes ) ) > key_size )
+        return( PSA_ERROR_BUFFER_TOO_SMALL );
+
+    /* Generate key data. */
+    if( PSA_KEY_TYPE_IS_UNSTRUCTURED( attributes->core.type ) )
+    {
+        key_buffer_length = PSA_BITS_TO_BYTES( psa_get_key_bits( attributes ) );
+        status = psa_generate_random( key_buffer, key_buffer_length );
+        if( status != PSA_SUCCESS )
+            return( status );
+    }
+    else
+#if defined(MBEDTLS_ECP_C)
+    if ( PSA_KEY_TYPE_IS_ECC( attributes->core.type ) && PSA_KEY_TYPE_IS_KEY_PAIR( attributes->core.type ) )
+    {
+        psa_ecc_family_t curve = PSA_KEY_TYPE_ECC_GET_FAMILY( attributes->core.type );
+        mbedtls_ecp_group_id grp_id =
+            mbedtls_ecc_group_of_psa( curve, PSA_BITS_TO_BYTES( attributes->core.bits ) );
+        const mbedtls_ecp_curve_info *curve_info =
+            mbedtls_ecp_curve_info_from_grp_id( grp_id );
+        mbedtls_ecp_keypair ecp;
+        mbedtls_test_rnd_pseudo_info rnd_info;
+        memset( &rnd_info, 0x5A, sizeof( mbedtls_test_rnd_pseudo_info ) );
+
+        int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+        if( attributes->domain_parameters_size != 0 )
+            return( PSA_ERROR_NOT_SUPPORTED );
```

Comment:
Just ignore domain parameters. They aren't supported yet in Mbed TLS and not standardized in PSA.